saleor/checkout/fetch.py

```diff
@@ -50,6 +65,157 @@ def get_customer_email(self) -> str:
         return self.user.email if self.user else self.checkout.email
 
 
+@dataclass(frozen=True)
+class EmptyDeliveryMethod:
+    delivery_method: Optional[Union["ShippingMethod", "Warehouse"]] = None
+    shipping_address: Optional["Address"] = None
+    order_key: str = "shipping_method"
+
+    @property
+    def warehouse_pk(self) -> Optional[str]:
+        pass
+
+    @property
+    def is_local_collection_point(self) -> bool:
+        return False
+
+    @property
+    def delivery_method_name(self) -> Dict[str, Optional[str]]:
+        return {"shipping_method_name": None}
+
+    def get_warehouse_filter_lookup(self) -> Dict[str, Any]:
+        return {}
+
+    def calculate_checkout_shipping(
+        self, checkout_info: "CheckoutInfo", lines=None
+    ) -> TaxedMoney:
+        return zero_taxed_money(checkout_info.checkout.currency)
+
+    def is_valid_delivery_method(self) -> bool:
+        return False
+
+    def is_method_in_valid_methods(self, checkout_info: "CheckoutInfo") -> bool:
+        return False
+
+    def update_channel_listings(self, checkout_info: "CheckoutInfo") -> None:
+        checkout_info.shipping_method_channel_listings = None
+
+
+@dataclass(frozen=True)
+class ShippingMethodInfo(EmptyDeliveryMethod):
+    delivery_method: "ShippingMethod"
+    shipping_address: Optional["Address"]
+    order_key: str = "shipping_method"
+
+    @property
+    def delivery_method_name(self) -> Dict[str, Optional[str]]:
+        return {"shipping_method_name": smart_text(self.delivery_method)}
+
+    def calculate_checkout_shipping(self, checkout_info, lines=None) -> TaxedMoney:
```

Comment:
please note the circular dependency:  
`if lines is not None and all(isinstance(line, CheckoutLineInfo) for line in lines):`
We can't even import this method to `fetch.py`, This is the major reason why I have decided to put that code here. What is more the method of shipping calculation depends on shipping method, that is why i have decided to leverage polymorphism here. I have tried even different approach - injecting proper calculation method in DeliveryMethod builder, however due to two-way dependency this was really a daunting task :(

Reply:
Because, how we calculate shipping price depends on delivery method. Should I write another `calculate_shipping_method_for_c&c` or perform `if / else` inside base_checkout shipping price? If another `deliveryMethod` comes in, will we continue `if/else` scenario?  (Also, this method was hardcoded in AppManager)

Reply:
base_calculation should have enough details to calculate prices. You can wrap logic to calculate shipping_price for shipping method in helper function and call it in `if/else` in `base_calculations.shipping_price` (like you said). `ShippingMethodInfo` should only "tell" a base_calculation that the given delivery method is a shipping_method then base_caluclation calculates prices in a proper way.

Reply:
This is exactly the thing I want to avoid, but ok, I will do that.

Reply:
Done.